temporalio/worker/_workflow_instance.py

```diff
@@ -490,6 +539,7 @@ async def run_update() -> None:
                         f"Update handler for '{job.name}' expected but not found, and there is no dynamic handler. "
                         f"known updates: [{' '.join(known_updates)}]"
                     )
+                self._in_progress_updates[job.id] = (job, defn)
```

Comment:
As things stand, it has to be here, since `defn` is bound inside the `try`. I agree that the `try` should be scoped more tightly around the code which is actually raising the exceptions it's catching, but I think I'll leave that change out of this PR.